src/core/text-buffer.cc

```diff
@@ -378,6 +383,138 @@ struct TextBuffer::Layer {
     return result;
   }
 
+  struct SubsequenceMatchVariant {
+    size_t query_index = 0;
+    std::vector<uint32_t> match_indices;
+    int16_t score = 0;
+  };
+
+  vector<SubsequenceMatch> find_words_with_subsequence(u16string query, const u16string &extra_word_characters) {
+    size_t query_index = 0;
+    Point position;
+    Point current_word_start;
+    u16string current_word;
+
+    std::transform(query.begin(), query.end(), query.begin(), std::towlower);
+
+    // First, find the start position of all words matching the given
+    // subsequence.
+    std::unordered_map<u16string, vector<Point>> substring_matches;
+
+    for_each_chunk_in_range(Point(), extent(), [&] (TextSlice chunk) -> bool {
+      for (uint16_t c : chunk) {
+        bool is_word_character =
+          std::iswalnum(c) ||
+          std::find(extra_word_characters.begin(), extra_word_characters.end(), c) != extra_word_characters.end();
+
+        if (is_word_character) {
+          if (current_word.empty()) current_word_start = position;
+          current_word += c;
+          if (query_index < query.size() && towlower(c) == query[query_index]) {
+            query_index++;
+          }
+        } else {
+          if (!current_word.empty()) {
+            if (query_index == query.size()) {
+              substring_matches[current_word].push_back(current_word_start);
+            }
+            query_index = 0;
+            current_word.clear();
+          }
+        }
+
+        if (c == '\n') {
+          position.row++;
+          position.column = 0;
+        } else {
+          position.column++;
+        }
+      }
+
+      return false;
+    });
+
+    if (!current_word.empty() && query_index == query.size()) {
+      substring_matches[current_word].push_back(current_word_start);
+    }
+
+    // Next, calculate a score for each word indicating the quality of the
+    // match against the query.
+
+    static const unsigned consecutive_bonus = 5;
+    static const unsigned subword_start_bonus = 10;
+    static const unsigned mismatch_penalty = 1;
+    static const unsigned leading_mismatch_penalty = 3;
+
+    vector<SubsequenceMatch> matches;
+
+    for (auto entry : substring_matches) {
```

Comment:
I had to bust out a pen and paper to wrap my head around this algorithm, but it seems to work. You may want to think about this as some function that takes a word and a query as input and returns a score and match indices as output. Then, any number of scoring functions with the same type signature can be plugged-in.

```Word -> Query -> ([MatchIndex], Score)```

Reply:
Presumably you get the part where we explore every possible permutation of the query matching against the word. We could definitely extract a function if you'd like. In general our style has drifted toward waiting for a second use before extracting logic. While extracting can aid clarity locally, it also reduces clarity globally because now you have to reason about whether the extracted function has multiple callers.